include/shell.h

```diff
@@ -66,14 +66,16 @@ class HostShell {
 class BatchFile {
 public:
 	BatchFile(DOS_Shell* host, std::unique_ptr<ByteReader> input_reader,
-	          const char* const entered_name, const char* const cmd_line);
+	          const char* const entered_name, const char* const cmd_line,
+	          bool echo_on);
 	BatchFile(const BatchFile&)            = delete; // prevent copying
 	BatchFile& operator=(const BatchFile&) = delete; // prevent assignment
-	virtual ~BatchFile();
+	virtual ~BatchFile()                   = default;
 	virtual bool ReadLine(char* line);
 	bool Goto(std::string_view label);
 	void Shift();
-	bool echo                        = false;
+	void SetEcho(bool echo_on);
```

Comment:
> Previously, BatchFiles contained their parent's echo value

Do we know how real DOS behaves wrt batch files inheriting the parent shell's echo state?

Reply:
very good to know, thanks for digging into that!